nixos/modules/misc/nixpkgs.nix

```diff
@@ -324,74 +324,147 @@ in
+      };
+
+      assertions = let
+        # Whether `pkgs` was constructed by this module. This is false when any of
+        # nixpkgs.pkgs or _module.args.pkgs is set.
+        constructedByMe =
+          # We set it with default priority and it can not be merged, so if the
+          # pkgs module argument has that priority, it's from us.
+          (lib.modules.mergeAttrDefinitionsWithPrio options._module.args).pkgs.highestPrio
+            == lib.modules.defaultOverridePriority
+          # Although, if nixpkgs.pkgs is set, we did forward it, but we did not construct it.
+            && !opt.pkgs.isDefined;
+      in [
+        (
+          let
+            nixosExpectedSystem =
+              if config.nixpkgs.crossSystem != null
+              then config.nixpkgs.crossSystem.system or (lib.systems.parse.doubleFromSystem (lib.systems.parse.mkSystemFromString config.nixpkgs.crossSystem.config))
+              else config.nixpkgs.localSystem.system or (lib.systems.parse.doubleFromSystem (lib.systems.parse.mkSystemFromString config.nixpkgs.localSystem.config));
+            nixosOption =
+              if config.nixpkgs.crossSystem != null
+              then "nixpkgs.crossSystem"
+              else "nixpkgs.localSystem";
+            pkgsSystem = finalPkgs.stdenv.targetPlatform.system;
+          in {
+            assertion = constructedByMe -> !hasPlatform -> nixosExpectedSystem == pkgsSystem;
+            message = "The NixOS nixpkgs.pkgs option was set to a Nixpkgs invocation that compiles to target system ${pkgsSystem} but NixOS was configured for system ${nixosExpectedSystem} via NixOS option ${nixosOption}. The NixOS system settings must match the Nixpkgs target system.";
+          }
+        )
+        {
+          assertion = constructedByMe -> hasPlatform -> legacyOptionsDefined == [];
+          message = ''
+            Your system configures nixpkgs with the platform parameter${optionalString hasBuildPlatform "s"}:
+            ${hostPlatformLine
+            }${buildPlatformLine
+            }
+            However, it also defines the legacy options:
+            ${concatMapStrings showOptionWithDefLocs legacyOptionsDefined}
+            For a future proof system configuration, we recommend to remove
+            the legacy definitions.
+          '';
+        }
+        {
+          assertion = opt.pkgs.isDefined -> cfg.config == {};
+          message = ''
+            Your system configures nixpkgs with an externally created instance.
+            `nixpkgs.config` options should be passed when creating the instance instead.
+
+            Current value:
+            ${lib.generators.toPretty { multiline = true; } opt.config}
+          '';
+        }
+      ];
+    }
+    (mkIf (cfg.flake.source != null) (mkMerge [
+      (mkIf cfg.flake.setNixPath {
+        system.inputs.nixpkgs = mkDefault cfg.flake.source;
+        nix.nixPath = mkDefault [ "nixpkgs=/run/current-system/inputs/nixpkgs" ];
```

Comment:
I've been using `nix.nixPath = [ "nixpkgs=flake:nixpkgs" ]` for a while now, which makes the old CLI look up nixpkgs in the flake registry (since nix 2.13). And then I pin nixpkgs in the flake registry in the same way as you do here.

This avoids needing to create the additional symlink in the top-level derivation. I am not aware of any downsides.

Reply:
One downside is ease of inspection but honestly this is a really nice approach. I'll probably use it when I do a little rewriting of this one later once the lib changes are reviewed.